pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/util/HelixSetupUtils.java

```diff
@@ -175,4 +185,40 @@ private static HelixManager startHelixControllerInStandadloneMode(String helixCl
     return HelixControllerMain
         .startHelixController(zkUrl, helixClusterName, pinotControllerInstanceId, HelixControllerMain.STANDALONE);
   }
+
+  private static void createLeadControllerResourceIfNeeded(String helixClusterName, HelixAdmin admin,
+      boolean enableBatchMessageMode, boolean enableLeadControllerResource) {
+    List<String> resources = admin.getResourcesInCluster(helixClusterName);
+    if (!resources.contains(CommonConstants.Helix.LEAD_CONTROLLER_RESOURCE_NAME)) {
+      LOGGER.info("Cluster {} doesn't contain {}. Creating one..", helixClusterName,
+          CommonConstants.Helix.LEAD_CONTROLLER_RESOURCE_NAME);
+
+      admin.addStateModelDef(helixClusterName, CommonConstants.Helix.MASTER_SLAVE_STATE_MODEL_DEFINITION,
+          MasterSlaveSMD.build());
+
+      HelixHelper.updateResourceConfigsFor(new HashMap<>(), CommonConstants.Helix.LEAD_CONTROLLER_RESOURCE_NAME,
+          helixClusterName, admin);
+      admin.addResource(helixClusterName, CommonConstants.Helix.LEAD_CONTROLLER_RESOURCE_NAME,
+          CommonConstants.Helix.DEFAULT_NUMBER_OF_PARTITIONS_IN_LEAD_CONTROLLER_RESOURCE,
+          CommonConstants.Helix.MASTER_SLAVE_STATE_MODEL_DEFINITION, IdealState.RebalanceMode.FULL_AUTO.toString());
+
+      // Set instance group tag for lead controller resource.
+      IdealState leadControllerIdealState =
+          admin.getResourceIdealState(helixClusterName, CommonConstants.Helix.LEAD_CONTROLLER_RESOURCE_NAME);
+      leadControllerIdealState.setInstanceGroupTag(CommonConstants.Helix.CONTROLLER_INSTANCE_TYPE);
+      leadControllerIdealState.setBatchMessageMode(enableBatchMessageMode);
+      admin.setResourceIdealState(helixClusterName, CommonConstants.Helix.LEAD_CONTROLLER_RESOURCE_NAME,
+          leadControllerIdealState);
+    }
+
+    LOGGER.info("Re-balance lead controller resource with replicas: {}",
+        CommonConstants.Helix.DEFAULT_NUMBER_OF_CONTROLLER_REPLICAS);
+    admin.rebalance(helixClusterName, CommonConstants.Helix.LEAD_CONTROLLER_RESOURCE_NAME,
```

Comment:
Do we need to do this each time helix (or dual) controller is restarted? Only the first time right?

Reply:
Moved it inside of the bracket.